mod-source-record-manager-server/src/main/java/org/folio/services/EventDrivenChunkProcessingServiceImpl.java

```diff
@@ -25,29 +29,52 @@
                                                @Autowired JobExecutionService jobExecutionService,
                                                @Autowired ChangeEngineService changeEngineService,
-                                               @Autowired JobExecutionProgressService jobExecutionProgressService) {
+                                               @Autowired JobExecutionProgressService jobExecutionProgressService,
+                                               @Autowired MappingMetadataService mappingMetadataService) {
     super(jobExecutionSourceChunkDao, jobExecutionService);
     this.changeEngineService = changeEngineService;
     this.jobExecutionProgressService = jobExecutionProgressService;
+    this.mappingMetadataService = mappingMetadataService;
   }
 
   @Override
   protected Future<Boolean> processRawRecordsChunk(RawRecordsDto incomingChunk, JobExecutionSourceChunk sourceChunk, String jobExecutionId, OkapiConnectionParams params) {
     LOGGER.debug("Starting to process raw records chunk for jobExecutionId: {}", jobExecutionId);
     Promise<Boolean> promise = Promise.promise();
-
     initializeJobExecutionProgressIfNecessary(jobExecutionId, incomingChunk, params.getTenantId())
       .compose(ar -> checkAndUpdateJobExecutionStatusIfNecessary(jobExecutionId, new StatusDto().withStatus(StatusDto.Status.PARSING_IN_PROGRESS), params))
       .compose(jobExec -> changeEngineService.parseRawRecordsChunkForJobExecution(incomingChunk, jobExec, sourceChunk.getId(), params))
+      .compose(records -> saveMappingMetaDataSnapshot(jobExecutionId, params, records))
       .onComplete(sendEventsAr -> updateJobExecutionIfAllSourceChunksMarkedAsError(jobExecutionId, params)
         .onComplete(updateAr -> promise.handle(sendEventsAr.map(true))));
     return promise.future();
   }
 
+  private Future<Boolean> saveMappingMetaDataSnapshot(String jobExecutionId, OkapiConnectionParams okapiParams, List<Record> recordsList) {
+    if (CollectionUtils.isEmpty(recordsList)) {
+      return Future.succeededFuture(false);
+    }
+    Promise<Boolean> promise = Promise.promise();
+    mappingMetadataService.getMappingMetadataDto(jobExecutionId, okapiParams)
+      .onComplete(arMappingMetadata -> {
+        if (arMappingMetadata.failed() && arMappingMetadata.cause() instanceof NotFoundException) {
+          mappingMetadataService.saveMappingRulesSnapshot(jobExecutionId, recordsList.get(0).getRecordType().toString(), okapiParams.getTenantId())
+            .compose(arMappingRules -> mappingMetadataService.saveMappingParametersSnapshot(jobExecutionId, okapiParams))
+            .onSuccess(ar -> promise.complete(true))
+            .onFailure(promise::fail);
+        }
+        if (arMappingMetadata.failed() && !(arMappingMetadata.cause() instanceof NotFoundException)) {
+          promise.future().failed();
+        }
+      });
```

Comment:
```suggestion
      .onSuccess(v -> promise.complete(false))
      .onFailure(e -> {
        if (e instanceof NotFoundException) {
          Record.RecordType recordType = recordsList.get(0).getRecordType();
          recordType = recordType == Record.RecordType.MARC_HOLDING ? recordType : Record.RecordType.MARC_BIB;
          mappingMetadataService.saveMappingRulesSnapshot(jobExecutionId, recordType.toString(), okapiParams.getTenantId())
            .compose(arMappingRules -> mappingMetadataService.saveMappingParametersSnapshot(jobExecutionId, okapiParams))
            .onSuccess(ar -> promise.complete(true))
            .onFailure(promise::fail);
        }
        promise.fail(e);
      });
```
propose making refactoring in this way to avoid code hanging when snapshots already exist

Reply:
Sure Ruslan, it is done. Thank you